### PR TITLE
Improve page table support

### PR DESCRIPTION
**Pull request by sboeuf**
Adding huge pages support along with some simplification for walking the page table hierarchy.

### CODE REVIEW

**Comment by sboeuf** [comment]
@tjeznach Please have a look :)

File: drivers/iommu/riscv/io_pgtable.c [hunk 1]
```diff
@@ -20,55 +20,108 @@
...
+				   unsigned shift, bool root, size_t pgsize,
+				   unsigned long (*pd_alloc)(gfp_t), gfp_t gfp,
+				   bool *updated)
 {
-	/* TODO: merge dev/iopgtable */
-	pgd_t *pgd;
-	p4d_t *p4d;
-	pud_t *pud;
-	pmd_t *pmd;
+	pte_t *pte;
```
Comment (tjeznach) [inline drivers/iommu/riscv/io_pgtable.c:77]: You can treat all PTEs as PMDs and use pmd_xxx accessors (here and in riscv_iommu_pt_walk_free()), including pmd_leaf(). There will be no need for extra pte_leaf(), what should btw always return true if its pte_t.   For pte_xxx use pmd_pte() to remap pointer.

File: arch/riscv/include/asm/pgtable.h [hunk 2]
```diff
@@ -314,6 +314,11 @@ static inline int pte_huge(pte_t pte)
 	return pte_present(pte) && (pte_val(pte) & _PAGE_LEAF);
 }
 
+static inline int pte_leaf(pte_t pte)
```
Comment (tjeznach) [inline arch/riscv/include/asm/pgtable.h:317]: I'd suggest not adding this at all. See comment @ io_pgtable.c#77

**Review by tjeznach — COMMENTED** [review]
LGTM. Suggestion to avoid adding pte_leaf().

File: drivers/iommu/riscv/io_pgtable.c (hunk 1 above)
Comment (sboeuf) [reply drivers/iommu/riscv/io_pgtable.c:77]: Ok I'll rework accordingly.

Comment (sboeuf) [reply drivers/iommu/riscv/io_pgtable.c]: Done!

**Comment by sboeuf** [comment]
@tjeznach please let me know if the replacement of PTEs with PMDs looks like you expected.

**Review by tjeznach — APPROVED** [review]
Thanks. Looks good.
Note, very likely there will be more rework of the page alloc/free code to enable PTEs allocation from constrained physical memory region.

**Pull request merged**